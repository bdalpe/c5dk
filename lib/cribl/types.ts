export type Constructor<T, Arguments extends unknown[] = any[]> = new(...arguments_: Arguments) => T;
export type AbstractConstructor<T, Arguments extends unknown[] = any[]> = abstract new(...arguments_: Arguments) => T;

export type VariableType = 'expression' | 'number' | 'object' | 'array';

export interface VariableArguments {
	type: VariableType;
	name: string;
}

