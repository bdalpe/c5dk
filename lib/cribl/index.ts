import {Construct} from "constructs";

export interface CriblProps {
	/**
	 * The directory to output the Cribl configuration YAML manifests.
	 *
	 * @default - `CRIBL_OUTDIR`, if defined, otherwise `dist`
	 */
	outdir?: string;
}

/**
 * The base construct for all Cribl configuration.
 *
 * @example
 * ```typescript
 * const cribl = new Cribl();
 *
 * // add your config here using cribl as the
 * // const group = new Group(cribl, 'default', {});
 * // ...
 *
 * // call synth() to write the configuration to disk
 * cribl.synth();
 * ```
 */
export class Cribl extends Construct {
    public readonly outdir: string;

	constructor(props?: CriblProps) {
		super(undefined as any, '');
		this.outdir = props?.outdir ?? process.env.CRIBL_OUTDIR ?? 'dist';
	}
}

export * from "./objects";
