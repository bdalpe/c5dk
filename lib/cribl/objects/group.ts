import {Construct} from "constructs";
import {BaseConstruct} from "../base";

export interface GroupProps {}

export class Group extends BaseConstruct {
	constructor(scope: Construct, id: string, props: GroupProps) {
		super(scope, id, props);
	}
}
