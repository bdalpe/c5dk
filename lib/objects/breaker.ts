import {Construct} from "constructs";
import {ConfigConstruct} from "../config";
import Registry from "../registry";

export interface EventBreakerRulesetProps {
	lib: string;
	description?: string;
	tags?: string;
}

export class EventBreakerRuleset extends ConfigConstruct {
	constructor(scope: Construct, id: string, props: EventBreakerRulesetProps) {
		super(scope, id, props);
	}

	get kind() {
		return 'breakers';
	}

	get config(): Record<string, unknown> {
		return {...super.config, rules: this.node.children.filter(x => x instanceof EventBreakerRule).map(rule => rule.config)};
	}

	static dump(config: ConfigConstruct[]): Record<string, unknown> {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.register('breakers', EventBreakerRuleset);

export interface EventBreakerRuleProps {
	name: string;
	disabled: boolean;
	condition: string;
	type: string;
}

export class EventBreakerRule extends ConfigConstruct {
	constructor(scope: Construct, id: string, props: EventBreakerRuleProps) {
		// Rules can only be children of a ruleset
		if (!(scope instanceof EventBreakerRuleset)) {
			throw new Error('EventBreakerRule must be a child of EventBreakerRuleset');
		}

		super(scope, id, props);
	}

	get kind() {
		return 'rule';
	}
}
