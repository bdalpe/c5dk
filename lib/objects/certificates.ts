import {ConfigConstruct} from "../config";
import {pick} from "es-toolkit";
import {Construct} from "constructs";
import {File} from "../private/fs";
import Registry from "../registry";

export interface CertificateProps {
	/**
	 * Brief description of the certificate.
	 *
	 * Optional.
	 */
	description?: string;

	/**
	 * The certificate in PEM/Base64 format.
	 */
	certificate: string;

	/**
	 * The private key in PEM/Base64 format.
	 */
	privateKey: string;

	/**
	 * The passphrase for the private key.
	 *
	 * Optional.
	 */
	passphrase?: string;

	/**
	 * The CA certificate in PEM/Base64 format.
	 *
	 * Optional.
	 */
	caCert?: string;
}

export class Certificate extends ConfigConstruct {
	protected certificate: string;
	protected privateKey: string;
	protected passphrase : string | undefined;
	protected caCert: string | undefined;

	constructor(scope: Construct, id: string, props: CertificateProps) {
		super(scope, id, pick(props, ['description']));

		this.certificate = props.certificate;
		this.privateKey = props.privateKey;
		this.passphrase = props.passphrase;
		this.caCert = props.caCert;
	}

	get kind() {
		return 'certificates';
	}

	synth(): void {
		new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.pem`)).write(this.certificate);
		new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.key`)).write(this.privateKey);
		if (this.passphrase) new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.pas`)).write(this.passphrase);
		if (this.caCert) new File(this.path('local', 'cribl', 'auth', 'certs', `${this.node.id}.crt`)).write(this.caCert);
	}

	static dump(config: ConfigConstruct[]): Record<string, unknown> {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.register('certificates', Certificate);
