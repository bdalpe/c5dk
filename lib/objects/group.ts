import {ContainerConstruct} from "../container";
import {Construct} from "constructs";
import Registry from "../registry";

export interface GroupProps {
	description?: string;
}

export class WorkerGroup extends ContainerConstruct {
	constructor(scope: Construct, id: string, protected props?: GroupProps) {
		super(scope, id, props ?? {});
	}

	get kind() {
		return 'groups';
	}

	static dump(config: ContainerConstruct[]): Record<string, unknown> | undefined {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.registerContainer('groups', WorkerGroup)
