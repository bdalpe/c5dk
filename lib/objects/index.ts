export * from './output';
export * from './breaker';
export * from './certificates';
export * from './fleet';
export * from './license';
export * from './sample';
export * from './scripts';
export * from './variable';
export * from './regex'
