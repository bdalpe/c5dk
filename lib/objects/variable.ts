import {VariableArguments, VariableType} from "../types";
import Registry from "../registry";
import {ConfigConstruct} from "../config";

export interface VariableProps {
	/**
	 * The type of the variable.
	 */
	type: VariableType;

	/**
	 * The library to use for the variable.
	 *
	 * The only supported value is `cribl`.
	 *
	 * You may break things if using a non-supported value!
	 */
	lib?: string;

	/**
	 * The value of the variable.
	 */
	value: string;

	/**
	 * A description of the variable.
	 */
	description?: string;

	/**
	 * A list of arguments to pass to an expression.
	 */
	args?: Array<VariableArguments>;

	/**
	 * A list of tags to display in the UI.
	 */
	tags?: string[];
}

export class Variable extends ConfigConstruct {
	get kind() {
		return 'vars';
	}

	static dump(config: ConfigConstruct[]): Record<string, unknown> {
		return Object.fromEntries(config.map(c => [c.node.id, c.config]));
	}
}

Registry.register('vars', Variable);
