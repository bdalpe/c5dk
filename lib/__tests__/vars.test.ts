import { describe, it, expect, vi, beforeEach } from 'vitest';
import { fs, vol } from 'memfs'
import {Cribl} from "../cribl";
import {Variable} from "../cribl/objects/variable";
import {WorkerGroup} from "../cribl/objects/group";

vi.mock('node:fs');
vi.mock('node:fs/promises');

let cribl: Cribl;

beforeEach(() => {
  // reset the state of in-memory fs
  vol.reset()

	cribl = new Cribl({outdir: '/tmp'});
});

it('should write a file', () => {
	new Variable(cribl, 'test', {type: 'expression', value: 'test'});
	cribl.synth();

	const volume = vol.toJSON();
	expect(Object.keys(volume)).toEqual(['/tmp/local/cribl/vars.yml']);
	expect(volume['/tmp/local/cribl/vars.yml']).toMatchSnapshot();
})

it('should write a file with arguments', () => {
	new Variable(cribl, 'test', {type: 'expression', value: 'test', args: [{type: 'expression', name: 'test'}]});
	cribl.synth();

	const volume = vol.toJSON();
	expect(Object.keys(volume)).toEqual(['/tmp/local/cribl/vars.yml']);
	expect(volume['/tmp/local/cribl/vars.yml']).toMatchSnapshot();
})

it('should handle multiple variables', () => {
	new Variable(cribl, 'test', {type: 'expression', value: 'test'});
	new Variable(cribl, 'test2', {type: 'expression', value: 'test2'});
	cribl.synth();

	const volume = vol.toJSON();
	expect(Object.keys(volume)).toEqual(['/tmp/local/cribl/vars.yml']);
	expect(volume['/tmp/local/cribl/vars.yml']).toMatchSnapshot();
});

it('should be in a worker group', () => {
	const group = new WorkerGroup(cribl, 'goat')
	new Variable(group, 'test', {type: 'expression', value: 'test'});
	cribl.synth();

	const volume = vol.toJSON();
	expect(Object.keys(volume)).toEqual(['/tmp/groups/goat/local/cribl/vars.yml']);
	expect(volume['/tmp/groups/goat/local/cribl/vars.yml']).toMatchSnapshot();
});
