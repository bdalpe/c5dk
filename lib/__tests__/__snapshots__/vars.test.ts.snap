// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be in a worker group 1`] = `
"test:
  type: expression
  value: test
"
`;

exports[`should handle multiple variables 1`] = `
"test:
  type: expression
  value: test
test2:
  type: expression
  value: test2
"
`;

exports[`should write a file 1`] = `
"test:
  type: expression
  value: test
"
`;

exports[`should write a file with arguments 1`] = `
"test:
  type: expression
  value: test
  args:
    - type: expression
      name: test
"
`;
